dev/bots/test.dart

```diff
@@ -775,6 +780,72 @@ Future<void> _runWebIntegrationTests() async {
     'non-nullable',
     '--sound-null-safety',
   ]);
+  await _runGalleryE2eTest('debug');
+  await _runGalleryE2eTest('profile');
+  await _runGalleryE2eTest('release');
+  _chromeDriver?.kill();
+}
+
+Future<bool> _isChromeDriverRunning() async {
+  try {
+    (await Socket.connect('localhost', 4444)).destroy();
+    return true;
+  } on SocketException {
+    return false;
+  }
+}
+
+Future<void> _ensureChromeDriverIsRunning() async {
```

Comment:
Web installers package does this.

In february we discussed that this functionality will be needed in 3 repos. That was the reason we first started web_installers repo. I think we should keep using it instead of duplicating code.

Reply:
Happy to switch to web_installers when it's ready. Currently there's no way to add a dependency on web_installers. We could use a pub dependency, but web_installers is not on pub. We could git clone it, but that's not how flutter/flutter manages its pub dependencies and it's a lot more code to make the git cloning work smoothly.

Reply:
Why not, we are doing that since February? We are adding the dependency as SHA (as we are doing it with goldens).

https://github.com/flutter/engine/blob/master/lib/web_ui/pubspec.yaml#L35

(Actually going back to February, both creating web_installers and using it like this was your ides. You said you didn't want to create a copy code again (as we did for Chrome), so insisted that we should definitely have a web_installers repo for driver. I initially wanted to name that repo as "web_drivers" and you said Chrome should also be there. but never had a chance to carry Chrome :( )

Reply:
In short, the flutter/engine method doesn't work in flutter/flutter due to the single-version rule policy and avoidance of git dependencies (the `flutter update-packages --force-upgrade` does not understand them leading to issues).

If web_installers were in pub, we could use it here. I'm not sure if it's worth it though. All I'm trying to do is launch `chromedriver`, which is already in PATH, unless it's already running. If web_installers improves this over the code as written in the PR, I'm all for switching. We should be mindful, however, that this will add a dependency, which has costs, such as having to send multiple PRs and going through the `flutter update-packages --force-upgrade` process in order to fix bugs.

Reply:
As you mentioned in the offline discussion, if you believe "circumstances changed therefore we don't need a common repository" feel free to use this code. I personally favor common code usage. Especially considering this long running test will need:
- firefox driver code
- ios-safari simulator code
-safari driver code
in its lifetime, I feel like this class will be a very big copy/paste location. 

Note that if change something in the future (add verbose to drivers,  run on a different port), this is the only driver starting code outside of the `web_installers` common package. Therefore forgetting this will be easier. 

Also @godofredoc has a plan for separating tests/builds/dependency steps. When we do that,  this type of code (for example related to chrome driver or ios-simulator) will be in another step. When we look at a LUCI result, we will be able to clearly see "is tests failed", "is build failed", "is setup failed". 

I was hoping to use web_installers directly in that pre-steps. In this case looks like we would need to modify this class in the future.

If you want to use this code you can document these discussions to comments so if one webdriver tests fails and all the others are passing the developers who do the firefighting can clearly see how this one is different.

Reply:
Ah, I'm sorry, while checking the code I realize this part of the comment is not true: "Note that if change something in the future (add verbose to drivers, run on a different port), this is the only driver starting code outside of......". There are other places which use a stand alone chromedriver command.

Reply:
The risk of code duplication makes sense. Would you be able to publish `web_installers` to pub so we can reuse it across all repos like a regular pub dependency?

Reply:
Can you create an issue? I'm working on other P3 but I will try prioritize it.

Reply:
I will. In the meantime, can you PTAL at the PR again?

Reply:
Sure, I reviewed it all now :) Thanks for adding me as a reviewer.